proof/verifier.go

```diff
@@ -180,39 +180,56 @@ func (p *Proof) verifySplitRootProof() error {
 	return err
 }
 
-// verifyExclusionProofs verifies all ExclusionProofs are valid.
 func (p *Proof) verifyExclusionProofs() (*commitment.TapCommitmentVersion,
 	error) {
 
-	// Gather all P2TR outputs in the on-chain transaction.
-	p2trOutputs := make(map[uint32]struct{})
+	// There is nothing to check so return early.
+	if p.ExclusionProofs == nil {
+		return nil, nil
+	}
+
+	// Gather all P2TR outputs in the on-chain transaction. The STXO proofs
+	// are tracked using a set of serialized keys. We ignore that set when
+	// verifying V1 exclusion proofs.
+	p2trOutputs := make(map[uint32]fn.Set[asset.SerializedKey])
 	for i, txOut := range p.AnchorTx.TxOut {
 		if uint32(i) == p.InclusionProof.OutputIndex {
 			continue
 		}
-		if txscript.IsPayToTaproot(txOut.PkScript) {
-			p2trOutputs[uint32(i)] = struct{}{}
+		if !txscript.IsPayToTaproot(txOut.PkScript) {
+			continue
 		}
+
+		outIdx := uint32(i)
+		p2trOutputs[outIdx] = make(fn.Set[asset.SerializedKey])
 	}
 
-	// Verify all of the encoded exclusion proofs.
-	commitVersions := make(map[uint32]commitment.TapCommitmentVersion)
-	for _, exclusionProof := range p.ExclusionProofs {
-		exclusionProof := exclusionProof
-		derivedCommitment, err := verifyTaprootProof(
-			&p.AnchorTx, &exclusionProof, &p.Asset, false,
-		)
-		if err != nil {
-			return nil, err
-		}
+	// Early pass to determine if we're dealing with v1 or v2 proofs.
+	hasV2Proofs := len(p.ExclusionProofs) > 0 &&
+		p.ExclusionProofs[0].CommitmentProof != nil &&
+		len(p.ExclusionProofs[0].CommitmentProof.STXOProofs) > 0
+
+	if hasV2Proofs {
+		return p.verifyV2ExclusionProofs(p2trOutputs)
+	}
+
+	return p.verifyV1ExclusionProofs(p2trOutputs)
+}
+
+// verifyV1ExclusionProofs verifies all version 1 exclusion proofs.
+func (p *Proof) verifyV1ExclusionProofs(
+	p2trOutputs map[uint32]fn.Set[asset.SerializedKey],
+) (*commitment.TapCommitmentVersion, error) {
 
-		outputIdx := exclusionProof.OutputIndex
-		delete(p2trOutputs, outputIdx)
+	commitVersions := make(map[uint32][]commitment.TapCommitmentVersion)
 
-		// Store the commitment version. If there was no Taproot Asset
-		// commitment present, then there is nothing to store.
-		if derivedCommitment != nil {
-			commitVersions[outputIdx] = derivedCommitment.Version
+	// Verify all of the encoded v1 exclusion proofs.
+	for idx := range p.ExclusionProofs {
+		exclusionProof := p.ExclusionProofs[idx]
+		err := p.handleBasicExclusionProof(&exclusionProof,
```

Comment:
Style nit: Should make a new ilne after the first variable, close parens on the line after.